pkg/users/users.js

```diff
@@ -188,66 +177,76 @@ async function getLogins(logind_client) {
+                        details[name] = { ...details[name], isLocked: hash.startsWith("!") };
+                }
+            } catch (err) {
+                console.warn("Unexpected error when getting locked accounts from /etc/shadow:", err);
+            }
+        } else {
             console.warn("Unexpected error when getting locked account information", err);
         }
     }
 
-    let lastlog = "";
+    // last logged in
+
+    let LastLogPath;
     try {
-        lastlog = await cockpit.spawn([LastLogPath], { environ: ["LC_ALL=C"] });
-    } catch (ex) {
-        console.warn(`Failed to run ${LastLogPath}: ${ex.toString()}`);
+        await cockpit.spawn(["test", "-e", "/var/lib/lastlog/lastlog2.db"], { err: "ignore" });
+        LastLogPath = "lastlog2";
+    } catch (err1) {
+        LastLogPath = "lastlog";
     }
 
-    // drop header and last empty line with slice
-    const promises = lastlog.split('\n').slice(1, -1).map(async line => {
-        const splitLine = line.trim().split(/[ \t]+/);
-        const name = splitLine[0];
-        // Fallback on passwd -S for Fedora and RHEL
-        const isLocked = locked_users_map[name] ?? await get_locked(name);
-
-        if (line.indexOf('**Never logged in**') > -1) {
-            return { name, loggedIn: false, lastLogin: null, isLocked };
-        }
-
-        const loggedIn = currentLogins.includes(name);
-
-        const date_fields = splitLine.slice(-5);
-        // this is impossible to parse with Date() (e.g. Firefox does not work with all time zones), so call `date` to parse it
-        let lastLogin = null;
-        try {
-            const out = await cockpit.spawn(["date", "+%s", "-d", date_fields.join(' ')],
-                                            { environ: ["LC_ALL=C"], err: "out" });
-            lastLogin = parseInt(out) * 1000;
-        } catch (e) {
-            console.warn(`Failed to parse date from lastlog line '${line}': ${e.toString()}`);
-        }
+    try {
+        const out = await cockpit.spawn([LastLogPath], { environ: ["LC_ALL=C"] });
+        await Promise.all(out.split('\n').slice(1, -1).map(async line => {
+            if (line.includes('**Never logged in**'))
+                return;
 
-        return { name, loggedIn, lastLogin, isLocked };
-    });
+            const splitLine = line.trim().split(/[ \t]+/);
+            const name = splitLine[0];
+            const date_fields = splitLine.slice(-5);
+            // this is impossible to parse with Date() (e.g. Firefox does not work with all time zones), so call `date` to parse it
+            try {
+                const out = await cockpit.spawn(["date", "+%s", "-d", date_fields.join(' ')],
+                                                { environ: ["LC_ALL=C"], err: "out" });
+                details[name] = { ...details[name], lastLogin: parseInt(out) * 1000 };
+            } catch (e) {
+                console.warn(`Failed to parse date from lastlog line '${line}': ${e.toString()}`);
```

Comment:
These 2 added lines are not executed by any test.

Reply:
Unreachable error handler.